cfgmgr/vrrpmgr.cpp

```diff
@@ -0,0 +1,506 @@
+{
+}
+
+bool VrrpMgr::setIntfArpAccept(const std::string &intf_alias, const bool arp_accept)
+{
+    stringstream cmd;
+    string res;
+
+    if (arp_accept)
+    {
+        cmd << ECHO_CMD << " 2 > /proc/sys/net/ipv4/conf/" << shellquote(intf_alias) << "/arp_announce && ";
+        cmd << ECHO_CMD << " 2 > /proc/sys/net/ipv4/conf/" << shellquote(intf_alias) << "/rp_filter && ";
+        cmd << ECHO_CMD << " 1 > /proc/sys/net/ipv4/conf/" << shellquote(intf_alias) << "/accept_local";
+    }
+    else
+    {
+        cmd << ECHO_CMD << " 0 > /proc/sys/net/ipv4/conf/" << shellquote(intf_alias) << "/arp_announce && ";
+        cmd << ECHO_CMD << " 0 > /proc/sys/net/ipv4/conf/" << shellquote(intf_alias) << "/rp_filter && ";
+        cmd << ECHO_CMD << " 0 > /proc/sys/net/ipv4/conf/" << shellquote(intf_alias) << "/accept_local";
+    }
+
+    try
+    {
+        EXEC_WITH_ERROR_THROW(cmd.str(), res);
+    }
+    catch (const std::exception &e)
+    {
+        SWSS_LOG_ERROR("Failed to set intf arp %s on interface[%s], retry. Runtime error: %s", arp_accept ? "accept" : "default", intf_alias.c_str(), e.what());
+        return false;
+    }
+
+    SWSS_LOG_INFO("Set vrrp arp %s on interface[%s]", arp_accept ? "accept" : "default", intf_alias.c_str());
+    return true;
+}
+
+bool VrrpMgr::setVrrpIntf(const std::string &intf_alias, const std::string &vrid, const bool is_ipv4, 
+    const std::set<IpPrefix> &vip_list, const std::string &admin_status)
+{
+    VrrpIntfConf vrrp_conf;
+    if (m_vrrpList.find(vrid) == m_vrrpList.end())
+    {
+        vrrp_conf.alias = intf_alias;
+    }
+    else
+    {
+        vrrp_conf = m_vrrpList[vrid];
+    }
+    auto &vrrp = is_ipv4 ? vrrp_conf.vrrp4 : vrrp_conf.vrrp6;
+    auto &vrrp_entry = is_ipv4 ? vrrp_conf.vrrp4_entry : vrrp_conf.vrrp6_entry;
+
+    // generate vmac
+    set<IpPrefix> vaild_vips;
+    auto is_ipv4_check = [is_ipv4](const IpPrefix &has_vip){ return has_vip.isV4() == is_ipv4; };
+    copy_if(vip_list.begin(), vip_list.end(), std::inserter(vaild_vips, vaild_vips.begin()), is_ipv4_check);
+    if (!vaild_vips.empty())
+    {
+        // add vrrp intf
+        if (!vrrp.isValid())
+        {
+            MacAddress vmac;
+            if (!parseVrrpMac(vrid, is_ipv4, vmac))
+            {
+                return false;
+            }
+            vrrp = VrrpIntf(intf_alias, vrid, is_ipv4, vmac.to_string());
+            if (!vrrp.isValid())
+            {
+                SWSS_LOG_WARN("parse new vrrp intf fail, intf: %s, vrid: %s, is ipv4:%d", intf_alias.c_str(), vrid.c_str(), is_ipv4);
+                return false;
+            }
+            if (!addVirtualInterface(intf_alias, vrrp.getVrrpName(), vmac, is_ipv4))
+            {
+                vrrp = VrrpIntf();
+                return false;
+            }
+        }
+    }
+    else
+    {
+        // del vrrp intf
+        if (vrrp.isValid())
+        {
+            if (!delVirtualInterface(intf_alias, vrrp.getVrrpName()))
```

Comment:
IMO, MAC Vlan interface removal needs to be avoided when the VIP list is empty as tit is getting removed as soon as the assigned VIPs are removed from the VRRP interface.